File: examples/07-aws-containers/go/glue_activity.go
// Run a long Glue ETL job from inside an Activity: submit, heartbeat the run id
// while polling, and return a structured ApplicationError on a bad terminal state.
//
// The Go port of glue_activity.java. The AWS Glue calls sit behind a small
// GlueAPI interface so the Temporal usage compiles without the AWS SDK; wire the
// real aws-sdk-go-v2 glue.Client to that interface in production.
package aws

import (
	"context"
	"fmt"
	"time"

	"go.temporal.io/sdk/activity"
	"go.temporal.io/sdk/temporal"
)

// GlueAPI is the slice of the AWS Glue client this Activity needs.
type GlueAPI interface {
	StartJobRun(ctx context.Context, jobName, inputS3URI string) (runID string, err error)
	JobRunState(ctx context.Context, jobName, runID string) (state string, errMsg string, err error)
}

// GlueActivities wraps a GlueAPI so the Activity stays testable.
type GlueActivities struct{ Glue GlueAPI }

// RunGlueJob submits a Glue run, heartbeats the run id while polling, and maps
// terminal states to success vs. a typed failure.
func (a *GlueActivities) RunGlueJob(ctx context.Context, jobName, inputS3URI string) (string, error) {
	runID, err := a.Glue.StartJobRun(ctx, jobName, inputS3URI)
	if err != nil {
		return "", err
	}

	for {
		activity.RecordHeartbeat(ctx, runID)
		state, errMsg, err := a.Glue.JobRunState(ctx, jobName, runID)
		if err != nil {
			return "", err
		}

		switch state {
		case "SUCCEEDED":
			return runID, nil
		case "FAILED", "TIMEOUT", "STOPPED":
			// A typed failure surfaces in the UI as GlueJobFailed, not a stack trace.
			return "", temporal.NewApplicationError(
				fmt.Sprintf("Glue job %s ended as %s: %s", jobName, state, errMsg),
				"GlueJobFailed",
			)
		}

		// Back off between polls so we don't hammer the Glue API and burn the
		// throttle limit before the job finishes.
		select {
		case <-ctx.Done():
			return "", ctx.Err()
		case <-time.After(15 * time.Second):
		}
	}
}

File: examples/07-aws-containers/python/glue_activity.py
"""Run a long Glue ETL job from inside an Activity: submit, heartbeat the run id
while polling, and raise a structured ApplicationError on a bad terminal state.

The Python port of glue_activity.java. boto3 is imported lazily so this module
still py_compiles where boto3 is absent; the Temporal-side code is what matters.
"""

import asyncio
from datetime import timedelta

from temporalio import activity
from temporalio.exceptions import ApplicationError

_TERMINAL_BAD = {"FAILED", "TIMEOUT", "STOPPED"}


def _glue_client():
    import boto3  # lazy: AWS SDK may be absent offline

    return boto3.client("glue")


@activity.defn
async def run_glue_job(job_name: str, input_s3_uri: str) -> str:
    glue = _glue_client()
    run_id = glue.start_job_run(
        JobName=job_name, Arguments={"--input": input_s3_uri}
    )["JobRunId"]

    while True:
        activity.heartbeat(run_id)
        state = glue.get_job_run(JobName=job_name, RunId=run_id)["JobRun"][
            "JobRunState"
        ]

        if state == "SUCCEEDED":
            return run_id
        if state in _TERMINAL_BAD:
            # A typed failure surfaces in the UI as GlueJobFailed, not a stack trace.
            raise ApplicationError(
                f"Glue job {job_name} ended as {state}", type="GlueJobFailed"
            )

        # Back off between polls so we don't hammer the Glue API and burn the
        # throttle limit before the job finishes.
        await asyncio.sleep(timedelta(seconds=15).total_seconds())
